### PR TITLE
utils.pwsh: Ignore Strawberry Perl default path

### PR DESCRIPTION
**Pull request by RytoEX**
### Description

The Strawberry Perl installation is poisoning all of our native Windows builds (notably, FreeType, curl, PCRE, and Qt6). For now, ignoring its default installation path is likely the easiest solution.

Also, force Qt6 Windows builds to use bundled 3rd party libraries when possible (e.g., ZLIB, libpng, libjpeg).

### Motivation and Context

Want CI builds to stop failing.

While I have PRs ready for tightening up the FreeType and curl builds (FreeType manages to succeed regardless, the curl PR seemed to work), PCRE was the next to fail. We may want to continue fixing them individually one by one, and then apply this for future cases, or perhaps we don't want this, and we'd rather fix them all individually. In any case, I'm opening this PR to provide us the option of which path to choose.

### How Has This Been Tested?

CI is the test.

### Types of changes

- Bug fix (non-breaking change which fixes an issue)

### Checklist:

- [x] My code has been run through [clang-format](https://github.com/obsproject/obs-studio/blob/master/.clang-format).
- [x] I have read the [**contributing** document](https://github.com/obsproject/obs-studio/blob/master/CONTRIBUTING.rst).
- [x] My code is not on the master branch.
- [ ] The code has been tested.
- [x] All commit messages are properly formatted and commits squashed where appropriate.
- [x] I have included updates to all appropriate documentation.

### CODE REVIEW

**Comment by PatTheMav** [comment]
I've just spent about an hour on the Windows runner poking around the environment variables and what probably happens is that Strawberry perl provides a proper CMake package as such the prefix path we specify is not even used anymore.

This also affects all other dependencies being built, so I think that this is the more appropriate fix:

https://github.com/PatTheMav/obs-deps/commit/1aa076bf874120b22210243963700be6fd0181c5

**Comment by RytoEX** [comment]
> what probably happens is that Strawberry perl provides a proper CMake package as such the prefix path we specify is not even used anymore.

There are no ZLIB nor libpng nor libjpeg items in our specified prefix path for native Windows deps builds. They are part of the cross-compiled builds, which are done in parallel and not available here. The Strawberry Perl version did not change between runner versions (though the CMake version did).

I'm fairly certain that the Strawberry Perl path is used by curl (and possibly others), at least for perl.exe, so I don't think excluding the entire directory is right. I also want to avoid PATH manipulation unless we're running on CI.

Furthermore, the Qt changes here are just explicitly setting what has otherwise been happening all along.

**Comment by PatTheMav** [comment]
> > what probably happens is that Strawberry perl provides a proper CMake package as such the prefix path we specify is not even used anymore.
> 
> There are no ZLIB nor libpng nor libjpeg items in our specified prefix path for native Windows deps builds. They are part of the cross-compiled builds, which are done in parallel and not available here. The Strawberry Perl version did not change between runner versions (though the CMake version did).
> 
> I'm fairly certain that the Strawberry Perl path is used by curl (and possibly others), at least for perl.exe, so I don't think excluding the entire directory is right. I also want to avoid PATH manipulation unless we're running on CI.
> 
> Furthermore, the Qt changes here are just explicitly setting what has otherwise been happening all along.

freetype (among others) picks up libpng, zlib, and other packages because they are available in the strawberry install. There might be others I haven't noticed. Globally removing it is therefore safer and keeps us from finding more surprises like that. We also should not use curl, but `Invoke-Webrequest`, so a system-wide curl should never be found.

**Comment by RytoEX** [comment]
> freetype (among others) picks up libpng, zlib, and other packages because they are available in the strawberry install. There might be others I haven't noticed. Globally removing it is therefore safer and keeps us from finding more surprises like that. We also should not use curl, but `Invoke-Webrequest`, so a system-wide curl should never be found.

Correct, but I'm fairly sure that FreeType is fixed with CMAKE_IGNORE_PREFIX_PATH. I can go retest locally and confirm this later.

> We also should not use curl, but `Invoke-Webrequest`, so a system-wide curl should never be found.

This has nothing to do with curl in the PowerShell scripts. I mean that the curl build itself looks for and finds perl.exe in its own CMakeLists.

**Comment by PatTheMav** [comment]
In that case we should download and provide perl or check if it is truly necessary for building (which it shouldn't be iirc).

Deps should effectively build in an isolated environment, which is why I even looked at "resetting" the path to ensure no such sudden interference in the future. But that can wait for a future PR.

Qt changes looked fine.

**Comment by RytoEX** [comment]
> In that case we should download and provide perl or check if it is truly necessary for building (which it shouldn't be iirc).

Perl is already provided by the runner in `C:\Strawberry\perl\bin\perl.exe`, just as it is on my machines where I have already installed Strawberry for other projects (such as building Qt). Strawberry Perl is _widely recommended_ as _the way_ to get Perl on Windows, so this is probably not uncommon.

> Deps should effectively build in an isolated environment, which is why I even looked at "resetting" the path to ensure no such sudden interference in the future. But that can wait for a future PR.
> 
> Qt changes looked fine.

Does this mean that we can punt the discussion on further/alternative mitigation approaches for later and figure out if this is merge-able as-is?

> I'm fairly sure that FreeType is fixed with CMAKE_IGNORE_PREFIX_PATH. I can go retest locally and confirm this later.

Following up on this, I did go back and confirm this on one of my dev machines. The FreeType build respects CMAKE_IGNORE_PREFIX_PATH properly.

The Qt6 build does not seem to respect CMAKE_IGNORE_PREFIX_PATH when invoked via their configure wrapper script around CMake, probably because of how they handle additional defines, though I did not dig into that too deeply. Simply explicitly enforcing the build options to prefer bundled libraries, which is what was already occurring anyway, works for Qt6. I've left the option in place as a reminder to try using it again later if/when either Qt recommends using CMake directly (they still recommend their configure CMake wrapper) or we decide to use CMake directly on Windows anyway. Qt5 is not affected, presumably because it does not use CMake.

Given that the version of Strawberry did not change between runner versions, this leads me to believe that this is a change between CMake 3.24 and 3.25, which was changed between runner versions. Local testing seems to confirm this, as Qt6 builds fine on master with CMake 3.24.2, but not with CMake 3.25.0. The Qt configure log also shows not using system libraries on CMake 3.24.2. Testing FreeType against CMake 3.24.2 confirms the same observation.

**Comment by PatTheMav** [comment]
> > In that case we should download and provide perl or check if it is truly necessary for building (which it shouldn't be iirc).
> 
> Perl is already provided by the runner in `C:\Strawberry\perl\bin\perl.exe`, just as it is on my machines where I have already installed Strawberry for other projects (such as building Qt). Strawberry Perl is _widely recommended_ as _the way_ to get Perl on Windows, so this is probably not uncommon.

Can be discussed off thread.

> > Deps should effectively build in an isolated environment, which is why I even looked at "resetting" the path to ensure no such sudden interference in the future. But that can wait for a future PR.
> > Qt changes looked fine.
> 
> Does this mean that we can punt the discussion on further/alternative mitigation approaches for later and figure out if this is merge-able as-is?

Yep. Good enough for now, quicker turnaround is more important for now.

> > I'm fairly sure that FreeType is fixed with CMAKE_IGNORE_PREFIX_PATH. I can go retest locally and confirm this later.
> 
> Following up on this, I did go back and confirm this on one of my dev machines. The FreeType build respects CMAKE_IGNORE_PREFIX_PATH properly.
> 
> The Qt6 build does not seem to respect CMAKE_IGNORE_PREFIX_PATH when invoked via their configure wrapper script around CMake, probably because of how they handle additional defines, though I did not dig into that too deeply. Simply explicitly enforcing the build options to prefer bundled libraries, which is what was already occurring anyway, works for Qt6. I've left the option in place as a reminder to try using it again later if/when either Qt recommends using CMake directly (they still recommend their configure CMake wrapper) or we decide to use CMake directly on Windows anyway. Qt5 is not affected, presumably because it does not use CMake.
> 
> Given that the version of Strawberry did not change between runner versions, this leads me to believe that this is a change between CMake 3.24 and 3.25, which was changed between runner versions. Local testing seems to confirm this, as Qt6 builds fine on master with CMake 3.24.2, but not with CMake 3.25.0. The Qt configure log also shows not using system libraries on CMake 3.24.2. Testing FreeType against CMake 3.24.2 confirms the same observation.

Yeah as discussed off-thread this was confirmed and will be fixed in CMake 3.25.1. Using `CMAKE_IGNORE_PATH` supposedly is the "safer" solution.

**Review by PatTheMav — APPROVED** [review]
Looks fine to me.

File: utils.pwsh/Setup-Target.ps1 [hunk 1]
```diff
@@ -48,6 +48,7 @@ function Setup-BuildParameters {
         '-G', $VisualStudioId
         "-DCMAKE_INSTALL_PREFIX=$($script:ConfigData.OutputPath)"
         "-DCMAKE_PREFIX_PATH=$($script:ConfigData.OutputPath)"
+        "-DCMAKE_IGNORE_PREFIX_PATH=C:\Strawberry\c"
```
[the remark this answers is not in the document]
Comment (RytoEX) [reply utils.pwsh/Setup-Target.ps1:51]: It can also occur on local machines, at least until CMake 3.25.1 is available.

**Pull request merged**